File: ToyLanguage3/src/Model/openRFile.java
package Model;

import javafx.util.Pair;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class openRFile implements IStmt {
    private String var_file_id;
    private String filename;

    public openRFile(String var_file_id, String filename) {
        this.var_file_id = var_file_id;
        this.filename = filename;
    }

    @Override
    public PrgState execute(PrgState state) throws MyException, IOException {
        MyIFileDictionary<Integer, Pair<String, BufferedReader>> dict = state.getFileTable();
        try {
            for (int i = 0; i < dict.size(); i++) {
                if (dict.lookup(i).getKey().equals(filename))
                    throw new MyException("Filename already exists!");
            }
            BufferedReader br = null;
            FileReader fr = null;
            try {
                fr = new FileReader(filename);
                br = new BufferedReader(fr);
                Pair<String, BufferedReader> p = new Pair<>(filename,br);
                dict.add(p);
                int lastKey =dict.getKey();
                MyIDictionary<String,Integer> symTbl = state.getSymTable();
                symTbl.add(var_file_id,lastKey);
            } catch (IOException e) {
                throw new IOException(e);
            }
        } catch (MyException ex) {
            throw new MyException("" + ex);
        }
        return  state;
    }

    @Override
    public String toString() {
        return "OpenFile(" +
                var_file_id +
                "," +'\"' +filename + '\"'+")";
    }
}

File: ToyLanguage7/ToyLanguage7/OpenRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace ToyLanguage7
{
    class OpenRFile : IStmt
    {
        private string var_file_id;
        private string filename;
        private KeyGenerator key;

        public OpenRFile(string var_file_id, string filename,KeyGenerator key)
        {
            this.var_file_id = var_file_id;
            this.filename = filename;
            this.key = key;
        
        }

        public PrgState execute(PrgState state)
        {
            MyIFileDictionary<int, KeyValuePair<string, StreamReader>> dict = state.getFileTable();
            try
            {
                for (int i = 0; i < dict.size(); i++)
                {
                    if (dict.lookup(i).Key==filename)
                        throw new MyException("Filename already exists!");
                }
                StreamReader br = null;
                FileStream fr = null;
                try
                {
                    fr = new FileStream(filename, FileMode.Open);
                    br = new StreamReader(fr);
                    KeyValuePair<string, StreamReader> p = new KeyValuePair<string, StreamReader>(filename, br);
                    int uniqKey = this.key.getGeneratedKey();
                    dict.add(uniqKey,p);
                    //int lastKey = dict.getKey();
                    MyIDictionary<string, int> symTbl = state.getSymTable();
                    symTbl.add(var_file_id, uniqKey);
                }
                catch (IOException e)
                {
                    throw new IOException(e.ToString());
                }
            }
            catch (MyException ex)
            {
                throw new MyException("" + ex);
            }
            return state;
        }

        public override string ToString()
        {
            return "OpenFile(" +
                    var_file_id +
                    "," + '\"' + filename + '\"' + ")";
        }
    }
}
